generator/src/lib.rs

```diff
@@ -36,25 +36,30 @@ use pest_meta::{optimizer, unwrap_or_report, validator};
 
 pub fn derive_parser(input: TokenStream, include_grammar: bool) -> TokenStream {
     let ast: DeriveInput = syn::parse2(input).unwrap();
-    let (name, generics, path) = parse_derive(ast);
-
-    let root = env::var("CARGO_MANIFEST_DIR").unwrap_or_else(|_| ".".into());
-    let path = Path::new(&root).join("src/").join(&path);
-    let file_name = match path.file_name() {
-        Some(file_name) => file_name,
-        None => panic!("grammar attribute should point to a file"),
-    };
-
-    let data = match read_file(&path) {
-        Ok(data) => data,
-        Err(error) => panic!("error opening {:?}: {}", file_name, error),
+    let (name, generics, content) = parse_derive(ast);
+
+    let (data, path) = match content {
+        GrammarSource::File(ref path) => {
+            let root = env::var("CARGO_MANIFEST_DIR").unwrap_or_else(|_| ".".into());
```

Comment:
I think this should work fine.

```suggestion
            let root = env::var("CARGO_MANIFEST_DIR").unwrap_or(".".into());
```